# EDA - Customer Lifetime Value Analysis (CLV)

CLTV = ((Average Order Value x Purchase Frequency) / Churn Rate) x Profit Margin.

# Import libraries

In [25]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Prepare

In [26]:
data = pd.read_csv('data/raw/data.csv' , encoding="ISO-8859-1")

In [27]:
# Clean the formatting
data.to_csv('data/raw/data.csv', index=False, encoding='utf-8')

In [28]:
data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


In [29]:
# The shape of data (rows and column)
data.shape

(541909, 8)

Data has 541909 rows and 8 columns

# Process

## Missing data

In [30]:
data.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Missing data
- `Description`: 1454
- `CustormerID`: 135080 or 24.93% missing value (135080 / 135080 * 100 = 24.93%)

The significance missing value of `CustomerID` could lead to data bias or noise. To keep this clean, we will remove them.

In [31]:
data = data[data['CustomerID'].isna() == False]

In [32]:
# Check again after drop the null

data.isna().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

## Duplicated Data

In [33]:
data.duplicated().sum()

np.int64(5225)

In [34]:
# Remove duplicate data

data.drop_duplicates(inplace=True)

## Converting Data type

In [35]:
data.dtypes

InvoiceNo          str
StockCode          str
Description        str
Quantity         int64
InvoiceDate        str
UnitPrice      float64
CustomerID     float64
Country            str
dtype: object

`InvoiceDate` is currently string. Need to convert into date. `CustomerID` suppose to be a whole number which mean integer

In [36]:
# Convert string to datetime
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

In [37]:
# Convert float to integer
data['CustomerID'] = data['CustomerID'].astype(int, errors = 'ignore')